demos/tensorflow-image-tagging/src/demoapp/tagging.clj

```diff
@@ -0,0 +1,72 @@
+;;
+;;       Copyright (c) 2017 Two Sigma Investments, LP.
+;;       All Rights Reserved
+;;
+;;       THIS IS UNPUBLISHED PROPRIETARY SOURCE CODE OF
+;;       Two Sigma Investments, LP.
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns demoapp.tagging
+  (:require [clojure.java.io :as io]
+            [clojure.java.shell :as sh]
+            [clojure.string :as str]
+            [clojure.tools.logging :as log])
+  (:import (java.io File)))
+
+(defn- directory-location
+  [dir-name]
+  (let [base-dir (or (System/getenv "MESOS_DIRECTORY") (System/getProperty "user.dir"))]
+    (str base-dir File/separator dir-name)))
+
+(defn- copy-uri-to-file
+  [uri file]
+  (with-open [in (io/input-stream uri)
+              out (io/output-stream file)]
+    (io/copy in out)))
+
+(let [image-tagging-lock (Object.)]
+  (defn- retrieve-image-tags
+    [classifier-file image-location predictions]
+    (let [result-atom (atom nil)]
```

Comment:
What does the `atom` do?

Reply:
We store the result and return it after completing the `locking` block.

Reply:
Could the atom not be avoided entirely?  Just return the `sh/sh` call directly and remove the atom.

Reply:
`locking` returns `nil`, it does not return the value of the `body` passed to it.

Source: https://github.com/clojure/clojure/blob/clojure-1.9.0-alpha14/src/clj/clojure/core.clj#L1637